File: Assets/3D Platformer Tutorial/Scripts/Player/ThirdPersonPlayerAnimation.cs
using UnityEngine;
using System.Collections;

[System.Serializable]
// By default loop all animations
// The jump animation is clamped and overrides all others
// This is the jet-pack controlled descent animation.
// we actually use this as a "got hit" animation
// We are in full control here - don't let any other animations play when we start
// Fade in run
// We fade out jumpland quick otherwise we get sliding feet
// Fade in walk
// We fade out jumpland realy quick otherwise we get sliding feet
// Fade out walk and run
// We fell down somewhere
// We are not falling down anymore
 // Wall jump animation is played without fade.
 // We are turning the character controller 180 degrees around when doing a wall jump so the animation accounts for that.
 // But we really have to make sure that the animation is in full control so 
 // that we don't do weird blends between 180 degree apart rotations
[UnityEngine.AddComponentMenu("Third Person Player/Third Person Player Animation")]
public partial class ThirdPersonPlayerAnimation : MonoBehaviour
{
    public float runSpeedScale;
    public float walkSpeedScale;
    public Animation anim;
    public virtual void Start()
    {
        this.anim = this.GetComponent<Animation>();
        this.anim.wrapMode = WrapMode.Loop;
        this.anim["run"].layer = -1;
        this.anim["walk"].layer = -1;
        this.anim["idle"].layer = -2;
        this.anim.SyncLayer(-1);
        this.anim["ledgefall"].layer = 9;
        this.anim["ledgefall"].wrapMode = WrapMode.Loop;
        this.anim["jump"].layer = 10;
        this.anim["jump"].wrapMode = WrapMode.ClampForever;
        this.anim["jumpfall"].layer = 10;
        this.anim["jumpfall"].wrapMode = WrapMode.ClampForever;
        this.anim["jetpackjump"].layer = 10;
        this.anim["jetpackjump"].wrapMode = WrapMode.ClampForever;
        this.anim["jumpland"].layer = 10;
        this.anim["jumpland"].wrapMode = WrapMode.Once;
        this.anim["walljump"].layer = 11;
        this.anim["walljump"].wrapMode = WrapMode.Once;
        this.anim["buttstomp"].speed = 0.15f;
        this.anim["buttstomp"].layer = 20;
        this.anim["buttstomp"].wrapMode = WrapMode.Once;
        AnimationState punch = this.anim["punch"];
        punch.wrapMode = WrapMode.Once;
        this.anim.Stop();
        this.anim.Play("idle");
    }

    public virtual void Update()
    {
        ThirdPersonController playerController = (ThirdPersonController) this.GetComponent(typeof(ThirdPersonController));
        float currentSpeed = playerController.GetSpeed();
        if (currentSpeed > playerController.walkSpeed)
        {
            this.anim.CrossFade("run");
            this.anim.Blend("jumpland", 0);
        }
        else
        {
            if (currentSpeed > 0.1f)
            {
                this.anim.CrossFade("walk");
                this.anim.Blend("jumpland", 0);
            }
            else
            {
                this.anim.Blend("walk", 0f, 0.3f);
                this.anim.Blend("run", 0f, 0.3f);
                this.anim.Blend("run", 0f, 0.3f);
            }
        }
        this.anim["run"].normalizedSpeed = this.runSpeedScale;
        this.anim["walk"].normalizedSpeed = this.walkSpeedScale;
        if (playerController.IsJumping())
        {
            if (playerController.IsControlledDescent())
            {
                this.anim.CrossFade("jetpackjump", 0.2f);
            }
            else
            {
                if (playerController.HasJumpReachedApex())
                {
                    this.anim.CrossFade("jumpfall", 0.2f);
                }
                else
                {
                    this.anim.CrossFade("jump", 0.2f);
                }
            }
        }
        else
        {
            if (!playerController.IsGroundedWithTimeout())
            {
                this.anim.CrossFade("ledgefall", 0.2f);
            }
            else
            {
                this.anim.Blend("ledgefall", 0f, 0.2f);
            }
        }
    }

    public virtual void DidLand()
    {
        this.anim.Play("jumpland");
    }

    public virtual void DidButtStomp()
    {
        this.anim.CrossFade("buttstomp", 0.1f);
        this.anim.CrossFadeQueued("jumpland", 0.2f);
    }

    public virtual IEnumerator Slam()
    {
        this.anim.CrossFade("buttstomp", 0.2f);
        ThirdPersonController playerController = (ThirdPersonController) this.GetComponent(typeof(ThirdPersonController));
        while (!playerController.IsGrounded())
        {
            yield return null;
        }
        this.anim.Blend("buttstomp", 0, 0);
    }

    public virtual void DidWallJump()
    {
        this.anim.Play("walljump");
    }

    public ThirdPersonPlayerAnimation()
    {
        this.runSpeedScale = 1f;
        this.walkSpeedScale = 1f;
    }

}

File: Assets/3D Platformer Tutorial/Scripts/Player/ThirdPersonPlayerAnimation.js
var runSpeedScale = 1.0;
var walkSpeedScale = 1.0;

var anim : Animation;

function Start ()
{
	anim = GetComponent.<Animation>();

	// By default loop all animations
	anim.wrapMode = WrapMode.Loop;

	anim["run"].layer = -1;
	anim["walk"].layer = -1;
	anim["idle"].layer = -2;
	anim.SyncLayer(-1);

	anim["ledgefall"].layer = 9;	
	anim["ledgefall"].wrapMode = WrapMode.Loop;


	// The jump animation is clamped and overrides all others
	anim["jump"].layer = 10;
	anim["jump"].wrapMode = WrapMode.ClampForever;

	anim["jumpfall"].layer = 10;	
	anim["jumpfall"].wrapMode = WrapMode.ClampForever;

	// This is the jet-pack controlled descent animation.
	anim["jetpackjump"].layer = 10;	
	anim["jetpackjump"].wrapMode = WrapMode.ClampForever;

	anim["jumpland"].layer = 10;	
	anim["jumpland"].wrapMode = WrapMode.Once;

	anim["walljump"].layer = 11;	
	anim["walljump"].wrapMode = WrapMode.Once;

	// we actually use this as a "got hit" animation
	anim["buttstomp"].speed = 0.15;
	anim["buttstomp"].layer = 20;
	anim["buttstomp"].wrapMode = WrapMode.Once;	
	var punch = anim["punch"];
	punch.wrapMode = WrapMode.Once;

	// We are in full control here - don't let any other animations play when we start
	anim.Stop();
	anim.Play("idle");
}

function Update ()
{
	var playerController : ThirdPersonController = GetComponent(ThirdPersonController);
	var currentSpeed = playerController.GetSpeed();

	// Fade in run
	if (currentSpeed > playerController.walkSpeed)
	{
		anim.CrossFade("run");
		// We fade out jumpland quick otherwise we get sliding feet
		anim.Blend("jumpland", 0);
	}
	// Fade in walk
	else if (currentSpeed > 0.1)
	{
		anim.CrossFade("walk");
		// We fade out jumpland realy quick otherwise we get sliding feet
		anim.Blend("jumpland", 0);
	}
	// Fade out walk and run
	else
	{
		anim.Blend("walk", 0.0, 0.3);
		anim.Blend("run", 0.0, 0.3);
		anim.Blend("run", 0.0, 0.3);
	}
	
	anim["run"].normalizedSpeed = runSpeedScale;
	anim["walk"].normalizedSpeed = walkSpeedScale;
	
	if (playerController.IsJumping ())
	{
		if (playerController.IsControlledDescent())
		{
			anim.CrossFade ("jetpackjump", 0.2);
		}
		else if (playerController.HasJumpReachedApex ())
		{
			anim.CrossFade ("jumpfall", 0.2);
		}
		else
		{
			anim.CrossFade ("jump", 0.2);
		}
	}
	// We fell down somewhere
	else if (!playerController.IsGroundedWithTimeout())
	{
		anim.CrossFade ("ledgefall", 0.2);
	}
	// We are not falling down anymore
	else
	{
		anim.Blend ("ledgefall", 0.0, 0.2);
	}
}

function DidLand () {
	anim.Play("jumpland");
}

function DidButtStomp () {
	anim.CrossFade("buttstomp", 0.1);
	anim.CrossFadeQueued("jumpland", 0.2);
}

function Slam () {
	anim.CrossFade("buttstomp", 0.2);
	var playerController : ThirdPersonController = GetComponent(ThirdPersonController);
	while(!playerController.IsGrounded())
	{
		yield;	
	}
	anim.Blend("buttstomp", 0, 0);
}


function DidWallJump ()
{
	// Wall jump animation is played without fade.
	// We are turning the character controller 180 degrees around when doing a wall jump so the animation accounts for that.
	// But we really have to make sure that the animation is in full control so 
	// that we don't do weird blends between 180 degree apart rotations
	anim.Play ("walljump");
}

@script AddComponentMenu ("Third Person Player/Third Person Player Animation")